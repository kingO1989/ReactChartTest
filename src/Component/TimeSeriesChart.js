import React, { useEffect, useState } from 'react';
import {
    Chart as ChartJS,
    CategoryScale,
    LinearScale,
    PointElement,
    LineElement,
    Title,
    Tooltip,
    Legend,
} from 'chart.js';
import { Line } from 'react-chartjs-2';


ChartJS.register(
    CategoryScale,
    LinearScale,
    PointElement,
    LineElement,
    Title,
    Tooltip,
    Legend
);







//const labels = ['January', 'February', 'March', 'April', 'May', 'June', 'July'];



export default function TimeSeriesChart({ generalData, company }) {

    //select first 5 as labels
    //
    const [TS, setTS] = useState();


    const options = {
        responsive: true,
        plugins: {
            legend: {
                position: 'top',
            },
            title: {
                display: true,
                text: company,
            },
        },
    };
    /*   labeldata:labels,
      open:labelsRelatedData_open,
      high:labelsRelatedData_high,
      low:labelsRelatedData_low,
      close:labelsRelatedData_close,
      volume:labelsRelatedData_volume ; */


    const data = {
        labels: generalData.labeldata,
        datasets: [
            {
                label: 'Open',
                data: generalData.open,//labels.map(() => faker.datatype.number({ min: 0, max: 1000 })),
                borderColor: 'rgb(255, 99, 132)',
                backgroundColor: 'rgba(255, 99, 132, 0.5)',
            },
            {
                label: 'High',
                data: generalData.high,//labels.map(() => faker.datatype.number({ min: 0, max: 1000 })),
                borderColor: 'rgb(99, 255, 132)',
                backgroundColor: 'rgba(99, 255, 132, 0.5)',
            },
            {
                label: 'Low',
                data: generalData.low,//labels.map(() => faker.datatype.number({ min: 0, max: 1000 })),
                borderColor: 'rgb(0, 199, 132)',
                backgroundColor: 'rgba(0, 199, 132, 0.5)',
            },
            {
                label: 'Close',
                data: generalData.close,//labels.map(() => faker.datatype.number({ min: 0, max: 1000 })),
                borderColor: 'rgb(132, 99, 255)',
                backgroundColor: 'rgba(132, 99, 255), 0.5)',
            },
            /*   {
                  label: 'Dataset 2',
                  data: labels.map(() => faker.datatype.number({ min: 0, max: 1000 })),
                  backgroundColor: 'rgba(53, 162, 235, 0.5)',
              }, */
        ],
    };
    // console.log(labelsRelatedData)

    return <Line options={options} data={data} />;
}
