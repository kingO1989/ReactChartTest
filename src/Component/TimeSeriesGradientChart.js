import React, { useRef, useEffect, useState } from 'react';
//import type { ChartData, ChartArea } from 'chart.js';
import {
    Chart as ChartJS,
    CategoryScale,
    LinearScale,
    PointElement,
    LineElement,
    Tooltip,
    Legend,
} from 'chart.js';
import { Chart } from 'react-chartjs-2';
import faker from '@faker-js/faker';


ChartJS.register(
    CategoryScale,
    LinearScale,
    PointElement,
    LineElement,
    Tooltip,
    Legend
);

const colors = [
    'red',
    'orange',
    'yellow',
    'lime',
    'green',
    'teal',
    'blue',
    'purple',
];

function createGradient(ctx, area) {



    const colorStart = faker.random.arrayElement(colors);
    const colorMid = faker.random.arrayElement(
        colors.filter(color => color !== colorStart)
    );
    const colorEnd = faker.random.arrayElement(
        colors.filter(color => color !== colorStart && color !== colorMid)
    );

    const gradient = ctx.createLinearGradient(0, area.bottom, 0, area.top);

    gradient.addColorStop(0, colorStart);
    gradient.addColorStop(0.5, colorMid);
    gradient.addColorStop(1, colorEnd);

    return gradient;
}

export default function TimeSeriesGradientChart({ volumeData, company }) {
    const chartRef = useRef(null);
    const [chartData, setChartData] = useState({
        datasets: [],
    });




    const data = {
        labels: volumeData.labeldata,
        datasets: [
            {
                label: `${company} Volume `,
                data: volumeData.volume,
            },

        ],
    };




    useEffect(() => {
        const chart = chartRef.current;

        if (!chart) {
            return;
        }

        const chartData = {
            ...data,
            datasets: data.datasets.map(dataset => ({
                ...dataset,
                borderColor: createGradient(chart.ctx, chart.chartArea),
            })),
        };

        setChartData(chartData);
        //return
    }, [company]);

    return <Chart ref={chartRef} type='line' data={chartData} />;
}
