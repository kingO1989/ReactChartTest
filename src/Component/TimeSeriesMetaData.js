import { useEffect, useState } from "react";

export default function TimeSeriesMetaData({ data }) {

    const [metadata, setMetadata] = useState()
    useEffect(() => {

        setMetadata(() => []);
        // console.log(data.data)
        if (data) {

            for (const [key, value] of Object.entries(data)) {
                // console.log(key, value);
                setMetadata((prev) => [...prev, { key: key, value: value }])
            }
        }
    }, [data])
    return (<>
        <h1>Meta Data Information</h1>
        {
            metadata ?
                <>
                    {metadata.map((data, key) => {

                        return <div key={key}>{data.key} {data.value}</div>
                    })}
                </>

                : ""
        }




    </>);
}