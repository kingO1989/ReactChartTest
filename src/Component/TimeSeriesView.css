.view_container {
  display: flex;
  border: 3px dashed pink;
}

.view_container .metadata {
  width: 30%;
}

.view_container .viewdata {
  width: 68%;
  height: 600px;
  border: 1px solid black;
}

.view_container .viewdata .lineplot {
  width: 100%;
  height: 600px;
  border: 1px solid black;
}

.view_container .viewdata .gradientplot {
}
