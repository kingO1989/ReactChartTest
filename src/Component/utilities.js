export async function fetchData(timeseries, company) {
    let data = await fetch(`https://www.alphavantage.co/query?function=${timeseries}&symbol=${company}&interval=5min&apikey=${process.env.REACT_APP_ALPHA_V_APIKEY}`)
    data = await data.json()

    return data;

}