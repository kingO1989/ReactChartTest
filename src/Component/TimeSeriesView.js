import { useEffect, useState } from 'react'
import './TimeSeriesView.css'
import TimeSeriesMetaData from './TimeSeriesMetaData'
import TimeSeriesChart from './TimeSeriesChart'
import TimeSeriesGradientChart from './TimeSeriesGradientChart'

export default function TimeSeriesView({ data, company }) {

    const [metaData, setMetaData] = useState()
    const [TS, setTS] = useState([])
    const [volumeData, setVolumeData] = useState([])

    useEffect(() => {


        if (data) {
            setMetaData(data["Meta Data"]);
            let prem_arr = [];
            let arr = [];
            for (const [key, value] of Object.entries(data)) {
                // console.log(key, value);
                prem_arr.push({ key: key, value: value })
                // setTS((prev) => [...prev, { key: key, value: value }])
            }
            console.log(prem_arr)
            for (const [key, value] of Object.entries(prem_arr[1].value)) {
                // console.log(key, value);
                arr.push({ key: key, value: value })
                // setTS((prev) => [...prev, { key: key, value: value }])
            }
            let labels = [];
            let labelsRelatedData_open = [];
            let labelsRelatedData_high = [];
            let labelsRelatedData_low = [];
            let labelsRelatedData_close = [];
            let labelsRelatedData_volume = [];

            arr.map((data) => {
                labels.push(data.key)
                labelsRelatedData_open.push(Number(data.value["1. open"]))
                labelsRelatedData_high.push(Number(data.value["2. high"]))
                labelsRelatedData_low.push(Number(data.value["3. low"]))
                labelsRelatedData_close.push(Number(data.value["4. close"]))
                labelsRelatedData_volume.push(Number(data.value["5. volume"]))
            });
            let generalData = {
                labeldata: labels,
                open: labelsRelatedData_open,
                high: labelsRelatedData_high,
                low: labelsRelatedData_low,
                close: labelsRelatedData_close,
                volume: labelsRelatedData_volume
            }

            let volumeData = {
                labeldata: labels,
                volume: labelsRelatedData_volume
            }
            setVolumeData(volumeData)
            setTS(generalData);



        }

    }, [data])


    return (<>
        <div className="view_container">
            {metaData ?
                (
                    <>
                        <div className='metadata'>
                            <TimeSeriesMetaData data={metaData} />
                        </div>
                        <div className='viewdata'>

                            <div className='lineplot'>
                                <TimeSeriesGradientChart
                                    volumeData={volumeData}
                                    company={company}

                                />
                            </div>
                            <div className='lineplot'>
                                <TimeSeriesChart
                                    generalData={TS}
                                    company={company} /></div>



                        </div>


                    </>


                )

                : ""}
        </div>
    </>)
}