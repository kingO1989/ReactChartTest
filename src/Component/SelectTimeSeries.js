import { useState, useEffect } from "react";
import TimeSeriesView from "./TimeSeriesView";
import { fetchData } from "./utilities";
import TimeSeriesChart from "./TimeSeriesChart";


export default function SelectTimeSeries() {

    const [viewData, setViewData] = useState("");
    const [TS, setTS] = useState("TIME_SERIES_INTRADAY");
    const [company, setCompany] = useState("IBM");


    async function onChangeHandlerTS(target) {


        let timeseries
        if (target.value === undefined)
            timeseries = "TIME_SERIES_INTRADAY"
        else
            timeseries = target.value;

        setTS(timeseries);

        /*  let results = await fetchData(timeseries)
         console.log(results)
         setViewData(results) */
    }

    async function onChangeHandlerCN(target) {


        let cm
        if (target.value === undefined)
            cm = "IBM"
        else
            cm = target.value;
        setCompany(cm)



    }

    async function clickHandler() {
        let results = await fetchData(TS, company)
        console.log(results)
        setViewData(results)


    }


    useEffect(() => {

        //fetchData();

    }, [])
    return (
        <>
            <select onChange={(e) => onChangeHandlerTS(e.target)} data_test="time_series">
                <option  >Select Time series </option>
                <option value="TIME_SERIES_INTRADAY" >Intra-day</option>
                <option value="TIME_SERIES_DAILY">Daily</option>
                <option value="TIME_SERIES_DAILY_ADJUSTED">Daily adjusted</option>
            </select>
            <select onChange={(e) => onChangeHandlerCN(e.target)} data_test="company">
                <option  >Select Company </option>
                <option value="AAPL"  >AAPL	Apple Inc</option>
                <option value="MSFT"  >Microsoft Corporation</option>
                <option value="AMZN"  >Amazon.com, Inc.</option>
                <option value="NVDA"  >NVIDIA Corporation</option>
                <option value="BRK.B"  >Berkshire Hathaway Inc.</option>
                <option value="GOOGL"  >Alphabet Inc.</option>
                <option value="TSLA"  >Tesla, Inc.</option>
                <option value="META"  >Meta Platforms, Inc.</option>
                <option value="LLY"  >Eli Lilly and Company</option>
                <option value="UNH"  >UnitedHealth Group Incorporated</option>
                <option value="TSM"  >Taiwan Semiconductor Manufacturing Company Limited</option>
                <option value="JPM"  >JPMorgan Chase & Co.</option>
                <option value="AVGO"  >Broadcom Inc.</option>
                <option value="WMT"  >Walmart Inc.</option>
                <option value="XOM"  >Exxon Mobil Corporation</option>
                <option value="V"  >Visa</option>
                <option value="IBM"  >International Business Machine</option>


            </select>

            <button onClick={clickHandler} data_test="btn"> Fetch</button>

            <div data-testid={"mockview"} className="selected_view">

                <TimeSeriesView data={viewData} company={company} />

            </div>
        </>
    );
} 