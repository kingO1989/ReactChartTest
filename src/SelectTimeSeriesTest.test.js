
import TestRenderer from 'react-test-renderer'; // ES6
import SelectTimeSeries from "./Component/SelectTimeSeries";
import * as utils from "./Component/utilities";

/*
tool used : react-test-renderer module / TestRenderer
purpose: test snapshots and simulate render changes
implementation:
1.use Testrenderer to create a non browser render instance
2.convert to json, once in json array format used methods like map,find,filter to traverse nodes
3.once node has been retrieved use props to find properties or run methods
4.note once an method is rerun , it will rerender the Testrenderer create method must be used to create an updated reference
*/


it("snap shot test", async () => {
    //create a render of the SelectTimeSeries component
    const SelectTimeSeriesRenderer = TestRenderer.create(
        <SelectTimeSeries />
    );
    //spy on the fetchData method as it is important to the  SelectTimeSeries component
    const spy = jest.spyOn(utils, 'fetchData');


    //get the json extract so we can perform snapshots and actions
    let STSJson = SelectTimeSeriesRenderer.toJSON();


    //create snapshot here
    expect(STSJson).toMatchSnapshot();



    let tsSelectTree = STSJson.find((t) => t.props.data_test === "time_series");

    let cmSelectTree = STSJson.find((t) => t.props.data_test === "company");
    //extract the button child from the List element rendered
    let btn = STSJson.find((t) => t.props.data_test === "btn");




    //execute the onclick method, this should trigger a rerenderd
    await TestRenderer.act(async () => {
        tsSelectTree.props.onChange({ target: { value: 'TIME_SERIES_INTRADAY' } })
        cmSelectTree.props.onChange({ target: { value: 'AAPL' } })
        await btn.props.onClick();
    });
    //Component must have update, we can confirm by taking another snap shot
    STSJson = SelectTimeSeriesRenderer.toJSON();
    expect(STSJson).toMatchSnapshot();

    //test to see if our view as updated
    let mockViewTree = STSJson.find((t) => t.type === "div" && t.props.className === "selected_view");

    console.log(mockViewTree.children);

    expect(spy).toHaveBeenCalledTimes(1)
    // expect(mockViewTree.children[0].children[0]).toBe("name");

    spy.mockRestore();
})



