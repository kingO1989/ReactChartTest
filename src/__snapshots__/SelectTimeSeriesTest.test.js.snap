// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snap shot test 1`] = `
Array [
  <select
    data_test="time_series"
    onChange={[Function]}
  >
    <option>
      Select Time series 
    </option>
    <option
      value="TIME_SERIES_INTRADAY"
    >
      Intra-day
    </option>
    <option
      value="TIME_SERIES_DAILY"
    >
      Daily
    </option>
    <option
      value="TIME_SERIES_DAILY_ADJUSTED"
    >
      Daily adjusted
    </option>
  </select>,
  <select
    data_test="company"
    onChange={[Function]}
  >
    <option>
      Select Company 
    </option>
    <option
      value="AAPL"
    >
      AAPL Apple Inc
    </option>
    <option
      value="MSFT"
    >
      Microsoft Corporation
    </option>
    <option
      value="AMZN"
    >
      Amazon.com, Inc.
    </option>
    <option
      value="NVDA"
    >
      NVIDIA Corporation
    </option>
    <option
      value="BRK.B"
    >
      Berkshire Hathaway Inc.
    </option>
    <option
      value="GOOGL"
    >
      Alphabet Inc.
    </option>
    <option
      value="TSLA"
    >
      Tesla, Inc.
    </option>
    <option
      value="META"
    >
      Meta Platforms, Inc.
    </option>
    <option
      value="LLY"
    >
      Eli Lilly and Company
    </option>
    <option
      value="UNH"
    >
      UnitedHealth Group Incorporated
    </option>
    <option
      value="TSM"
    >
      Taiwan Semiconductor Manufacturing Company Limited
    </option>
    <option
      value="JPM"
    >
      JPMorgan Chase & Co.
    </option>
    <option
      value="AVGO"
    >
      Broadcom Inc.
    </option>
    <option
      value="WMT"
    >
      Walmart Inc.
    </option>
    <option
      value="XOM"
    >
      Exxon Mobil Corporation
    </option>
    <option
      value="V"
    >
      Visa
    </option>
    <option
      value="IBM"
    >
      International Business Machine
    </option>
  </select>,
  <button
    data_test="btn"
    onClick={[Function]}
  >
     Fetch
  </button>,
  <div
    className="selected_view"
    data-testid="mockview"
  >
    <div
      className="view_container"
    />
  </div>,
]
`;

exports[`snap shot test 2`] = `
Array [
  <select
    data_test="time_series"
    onChange={[Function]}
  >
    <option>
      Select Time series 
    </option>
    <option
      value="TIME_SERIES_INTRADAY"
    >
      Intra-day
    </option>
    <option
      value="TIME_SERIES_DAILY"
    >
      Daily
    </option>
    <option
      value="TIME_SERIES_DAILY_ADJUSTED"
    >
      Daily adjusted
    </option>
  </select>,
  <select
    data_test="company"
    onChange={[Function]}
  >
    <option>
      Select Company 
    </option>
    <option
      value="AAPL"
    >
      AAPL Apple Inc
    </option>
    <option
      value="MSFT"
    >
      Microsoft Corporation
    </option>
    <option
      value="AMZN"
    >
      Amazon.com, Inc.
    </option>
    <option
      value="NVDA"
    >
      NVIDIA Corporation
    </option>
    <option
      value="BRK.B"
    >
      Berkshire Hathaway Inc.
    </option>
    <option
      value="GOOGL"
    >
      Alphabet Inc.
    </option>
    <option
      value="TSLA"
    >
      Tesla, Inc.
    </option>
    <option
      value="META"
    >
      Meta Platforms, Inc.
    </option>
    <option
      value="LLY"
    >
      Eli Lilly and Company
    </option>
    <option
      value="UNH"
    >
      UnitedHealth Group Incorporated
    </option>
    <option
      value="TSM"
    >
      Taiwan Semiconductor Manufacturing Company Limited
    </option>
    <option
      value="JPM"
    >
      JPMorgan Chase & Co.
    </option>
    <option
      value="AVGO"
    >
      Broadcom Inc.
    </option>
    <option
      value="WMT"
    >
      Walmart Inc.
    </option>
    <option
      value="XOM"
    >
      Exxon Mobil Corporation
    </option>
    <option
      value="V"
    >
      Visa
    </option>
    <option
      value="IBM"
    >
      International Business Machine
    </option>
  </select>,
  <button
    data_test="btn"
    onClick={[Function]}
  >
     Fetch
  </button>,
  <div
    className="selected_view"
    data-testid="mockview"
  >
    <div
      className="view_container"
    >
      <div
        className="metadata"
      >
        <h1>
          Meta Data Information
        </h1>
        <div>
          1. Information
           
          Intraday (5min) open, high, low, close prices and volume
        </div>
        <div>
          2. Symbol
           
          IBM
        </div>
        <div>
          3. Last Refreshed
           
          2023-12-04 19:55:00
        </div>
        <div>
          4. Interval
           
          5min
        </div>
        <div>
          5. Output Size
           
          Compact
        </div>
        <div>
          6. Time Zone
           
          US/Eastern
        </div>
      </div>
      <div
        className="viewdata"
      >
        <div
          className="lineplot"
        >
          <canvas
            height={150}
            role="img"
            width={300}
          />
        </div>
        <div
          className="lineplot"
        >
          <canvas
            height={150}
            role="img"
            width={300}
          />
        </div>
      </div>
    </div>
  </div>,
]
`;
